multisig-wallet/src/main.sw

```diff
@@ -147,11 +147,51 @@ impl MultiSignatureWallet for Contract {
         true
     }
 
-    /// Takes in transaction data and hashes it into a unique tx hash
-    /// Used for verification of message
-    fn get_transaction_hash(to: ContractId, value: u64, data: b256, nonce: u64) -> b256 {
-        // TODO: data > b256?
-        _get_transaction_hash(to, value, data, nonce, contract_id())
+    /// Transfers assets to outputs & contracts
+    ///
+    /// # Panics
+    ///
+    /// - When the constructor has not been called to initialize the contract
+    /// - When the balance of the asset being sent is less than the balance in the contract
+    /// - When the public key cannot be recovered from a signature
+    /// - When the recovered addresses are not in ascending order (0x1 < 0x2 < 0x3...)
+    /// - When the signer is not an owner
+    /// - When the total approval count is less than the required threshold for execution
+    fn transfer(to: Sender, asset_id: ContractId, value: u64, data: b256, signature1: B512, signature2: B512) -> bool {
+        require(storage.nonce != 0, Error::NotInitialized);
+        require(value <= this_balance(asset_id), Error::InsufficientAssetAmount);
+
+        let tx_hash = _get_transaction_hash(to, value, data, storage.nonce, contract_id());
+
+        // TODO: change to Vec<B512> once implemented and then iterate instead of hardcoding length
+        let signer1: b256 = match ec_recover_address(signature1, tx_hash) {
+            Result::Ok(address) => address.value, _ => revert(42), 
+        };
+
+        let signer2: b256 = match ec_recover_address(signature2, tx_hash) {
+            Result::Ok(address) => address.value, _ => revert(42), 
+        };
+
+        require(~b256::min() < signer1 && signer1 < signer2, Error::IncorrectSignerOrdering);
+
+        let signer1_weight = get::<u64>(signer1);
+        let signer2_weight = get::<u64>(signer2);
+
+        require(signer1_weight != 0 && signer2_weight != 0, Error::NotAnOwner);
+
+        // Hardcoded value, that passes the checks above, until the loop below is unblocked
+        let approval_count = signer1_weight + signer2_weight;
+
+        require(storage.threshold <= approval_count, Error::ApprovalThresholdNotReached);
+
+        storage.nonce = storage.nonce + 1;
+        log(tx_hash);
+
+        match to {
+            Sender::Address(address) => transfer_to_output(value, asset_id, address), Sender::ContractId(contract) => force_transfer(value, asset_id, contract), 
```

Comment:
Why do we need a force_transfer branch here?  Is it because not all contracts may have implemented some sort of transfer/receive function?

Reply:
There are two functions

1. transfer_to_output()
    - Is used to transfer to an `Address`
2. force_transfer()
    - Is used to transfer to a `ContractId`

They are equivalent but the name `force_transfer` might be confusing since `transfer_to_output` is also forceful and cannot be stopped. Both of those functions act in the same way as if I were to drop a letter into your mail box.

Reply:
To my knowledge, in Sway contracts do not have a `receive` function like in `Solidity`

Reply:
@matt-user ensures that the receiving contract cannot possibly execute any code upon receipt of the coins. This will work for any contract, unlike calling a `receive` function and forwarding coins for example, where the receiving contract may not even implement a `receive` function, or may choose to revert the call, etc...
Forwarding coins with a call can be done visa any function that doesn't explicitly forbid it, but with `force_transfer` we don't even need to call a function on a contract.

Reply:
Note that without an explicit fallback function, if you try to call a contract in Sway without a function selector (or with a selector that's not in the contract's ABI), the ABI dispatcher will _revert_. This means there's no much point in using call to transfer assets to a contract here. If you want to specifically call a function to receive funds, then you can call it.